README
HTTP::MHTTP
=============

    Copyright (c) 2003 Piers Harding.
    All rights reserved.
    This program is free software; you can redistribute it and/or modify
    it under the terms of either:
        a) the GNU General Public License as published by the Free
        Software Foundation; either version 1, or (at your option) any
        later version, or
        b) the "Artistic License"


HTTP::MHTTP is a fast http client library for Perl.  its advantages are that 
it gives you low level access to constructing the HTTP requests, and it 
handles GET, POST, PUT, HEAD, and DELETE.

Basic SSL support can be compiled in - this does not validate certificates,
but enables the TLS level negotiation.

To compile in SSL support then amend the call to perl Makefile.PL to be:
GOTSSL=yes perl Makefile.PL

You may need to amend the Makefile.PL script to ensure that the CFLAGS, and 
LDFLAGS (INC, and LIBS) are pointing at your openssl installation correctly.

The test suite assumes that you have access to the internet - so if the
tests fail then it is most likely due to you not being able to hit
http://www.piersharding.com

perl Makefile.PL   <= change this to GOTSSL=yes perl Makefile.PL for SSL support
make 
make test
make install

Dance.

for more details:

perldoc HTTP::MHTTP

Piers Harding - piers@cpan.org
Somewhere arround: Fri Jan  3 09:48:04 GMT 2003


Changes
Revision history for Perl module HTTP::MHTTP

0.16 ??
    - Fixed error in SYNOPSIS: was checking truth-value of http_call(),
      but should check whether the return is positive.
    - Added MIME::Base64 and Carp as pre-requisites
    - Reformatted the NAME section to give a brief one-line description
    - Reformatted this file according to CPAN::Changes::Spec

0.15 2003-12-14
    - Change the response headers to be an array ref for mhttp_split_headers()

0.14 2003-09-22
    - added -DOPENSSL_NO_KRB5 to suppress Kerberos header includes

0.13 2003-05-19
    - enhanced the build process to determine where openssl is installed -
      this was all borrowed from Crypt-SSLeay

0.11 2003-03-24
    - fix a bug with empty body,
      and move to get SSL validating certificates properly

0.10 2003-03-24
    - add in basic SSL support - this does not validate certificates,
      but enables the TLS level negotiation.

0.09 2003-03-02
    - fixed up the fudge for Solaris and herror() - I accidentally
      deleted it, and remove reference to timeval.

0.08 2003-02-28
    - Added basic HTTP 1.1, and Keep-Alive Support

0.07 2003-01-30
    - define our own herror() if we are on Solaris

0.06 2003-01-29
    - change the require to 5.005 - as tested by Mike Chamberlain

0.05 2003-01-25
    - fixed up a test that was pointing to localhost!

0.04 2003-01-23
    - modified the C library to realloc() space for the incoming buffers



Makefile.PL
use strict;
use ExtUtils::MakeMaker;
# most of this has been taken from Crypt-SSLeay


my $GOTSSL = "";
my($SEP, $SSL_LIB, $SSL_INC, $LIBS);

# FIND POSSIBLE SSL INSTALLATIONS
my @POSSIBLE_SSL_DIRS = qw(/usr/local/openssl /usr/local/ssl /local/ssl /opt/ssl /usr/local /local /usr);
my $open_ssl = 0;


# if they want to build with ssl then they will have set this env variable
if (exists $ENV{'GOTSSL'} && $ENV{'GOTSSL'} eq 'yes'){

  use File::Basename qw(dirname basename);
  use Getopt::Long;
  use vars qw($opt_default);

  &GetOptions("default", \$opt_default);

  $opt_default ||= $ENV{CRYPT_SSLEAY_DEFAULT};
  my $currdir = dirname($0) || '.';
  chdir($currdir) || die("can't change to $currdir: $!");
  if($^O eq 'MSWin32') {
      unshift(@POSSIBLE_SSL_DIRS, 'c:/openssl');
  }

  my @CANDIDATE;
  my $dir;
  for $dir (@POSSIBLE_SSL_DIRS) {
      next unless -d $dir;
      my $candidate = Candidate($dir);
      next unless $candidate;
  
      push(@CANDIDATE, $candidate);
  }

  if (@CANDIDATE == 1) {
      my($dir, $inc_dir, $ver) = @{$CANDIDATE[0]};
      print "Found OpenSSL (version $ver) installed at $dir\n";
  } elsif (@CANDIDATE > 1) {
      print "Found the following OpenSSL installations:\n";
      for (@CANDIDATE) {
  	my($dir, $inc_dir, $ver) = @$_;
  	print "\t$ver\t$dir\n";
      }
  }

  my $SSL_DIR;
  if($opt_default && (@CANDIDATE == 1) && $CANDIDATE[0][0]) {
      $SSL_DIR = $CANDIDATE[0][0];
      print "Using --default OpenSSL candidate found at $SSL_DIR\n";
  } else {
      if($ARGV[0] =~ /^[a-z]\:|^\//i) {
  	$SSL_DIR = shift;
      } else {
  	unless($CANDIDATE[0][0]) {
  	    print "No OpenSSL installation found, usually in $POSSIBLE_SSL_DIRS[0]\n";
  	}
  	$SSL_DIR = prompt "Which OpenSSL build path do you want to link against?", $CANDIDATE[0][0];
      }
  }

  my $candidate = &Candidate($SSL_DIR);
  unless($candidate) {
      warn "Apparently no SSLeay installation at '$SSL_DIR'\nAre you sure you got it correct????\n"
        unless -f "$SSL_DIR/include/ssl.h";
  }
  $candidate ||= ["$SSL_DIR", "$SSL_DIR/include"];
  
  $SSL_INC = "$SSL_DIR/include";
  
  if($^O =~ /Win32/) {
      unless($SSL_DIR =~ /^[a-zA-Z]\:/) {
  	# default drive c
  	$SSL_DIR = "c:".$SSL_DIR;
      }
      unless($SSL_INC =~ /^[a-zA-Z]\:/) {
  	# default drive c
  	$SSL_INC = "c:".$SSL_INC;
      }
      $SSL_INC .= " -I".$SSL_DIR."/inc32";
      $SSL_DIR =~ s|/|\\|g;
      $SSL_INC =~ s|/|\\|g;
  
      # patch from Ben Laurie
      if(-d "$SSL_DIR/lib") {
    	$SSL_LIB = "-L$SSL_DIR/lib";
      } elsif(-d "$SSL_DIR/out32dll") {
  	$SSL_LIB = "-L$SSL_DIR\\out32dll";	
      }else {
  	# Allow developers to point at OpenSSL source...
  	$SSL_LIB = "-L$SSL_DIR";
      }

      $SEP = "\\";
      $LIBS = "-lssleay32 -llibeay32";
  } else {
      $SSL_LIB = "-L".$SSL_DIR."/lib";
      $SEP = "/";
      $LIBS = "-lssl -lcrypto ".($^O ne 'linux' ? " -lgcc" : "");
  
      # ccc on alpha support
      if ($^O eq 'linux' && `uname -m` =~ /alpha/ &&
  	!(system("nm $SSL_DIR/lib/libssl.a|grep -q 'U _Ots'")>>8)) {
  	$LIBS .= ' -lots'; 
      }
  }

  # write include file that determing ssl support
  my $ssl_header_prefix = '';
  my $candidate_info = join('; ', @$candidate);
  if($candidate->[1] =~ /openssl/i) {
      $ssl_header_prefix = "openssl/";
  }
  
  print <<INFO;

================================================
BUILD INFORMATION
================================================

ssl dir:	$SSL_DIR
libraries:      $LIBS
include dir:	$SSL_INC
ssl header:     @{[$ssl_header_prefix]}ssl.h
ssl candidate:  $candidate_info

================================================

INFO
  ;

# end of Crypt-SSLeay

  $GOTSSL = ' -DOPENSSL_NO_KRB5 -DGOTSSL ';
}

WriteMakefile(
    clean => { 'FILES'	=> '*~ *.inl core test*.txt *.tar.gz _Inline ' },
    dist => { 'TARFLAGS'	=> 'cvf',
              'COMPRESS'        => 'gzip -9f',
    		'SUFFIX'	=> '.tgz'},
    'NAME'	=> 'HTTP::MHTTP',
    'VERSION_FROM' => 'MHTTP.pm',
    'MYEXTLIB' => 'mhttp/libmhttp$(LIB_EXT)',
    'DEFINE'	=> $GOTSSL,
    'LIBS'      => [($GOTSSL ? "$SSL_LIB $LIBS" : "")],
    'INC'	=> "-Imhttp ".($GOTSSL ? "-I".$SSL_INC : ""),
    PREREQ_PM => {
        'Carp'         => 0,
        'MIME::Base64' => 0,
    },
);

sub MY::postamble {
'
$(MYEXTLIB): mhttp/Makefile
	cd mhttp && $(MAKE) $(PASSTHRU)
';
}


# this next bit is taken from Crypt-SSLeay
## HELPERS

sub Candidate {
    my $dir = shift;
    my $version_file;
    my $inc_dir;

    for (
	 "$dir/crypto/opensslv.h", # cygwin32 builds
	 "$dir/inc32/openssl/opensslv.h", # win32 builds
	 "$dir/include/openssl/opensslv.h", 
	 "$dir/include/opensslv.h", 
	 "$dir/include/crypto.h"
	) 
      {
	  if(-e $_) {
	      $version_file = $_;
	  } 
      }
    return unless defined $version_file;

    $open_ssl = ($version_file =~ /openssl/) ? 1 : 0;
    my $dirname = $inc_dir = dirname($version_file);
    return unless (-e "$dirname/ssl.h");

    open(VERSION_FILE, $version_file) or return;
    my $version_match = $open_ssl ? "OPENSSL_VERSION_NUMBER" : "SSLEAY_VERSION_NUMBER";
    my $version;
    while (<VERSION_FILE>) {
	if (/^\#define\s+$version_match\s+0x0+(\d\d\d)/) {
	    $version = $1;
	    $version =~ s/(\d)0(\d)/$1$2/;	    
	    my $type = ($version > 92) ? "OpenSSL" : "SSLeay";
	    $version = "$type ".join('.', split(//, "0$version"));
	    last;
	}
    }
    close(VERSION_FILE);

    # Silly test to look for the library files
    my $foundlib = 0;
    my $libd = (-d "$dir/out32dll") ? 'out32dll' : 'lib';
    if (opendir(LIBDIR, "$dir/$libd")) {
	while (defined($_ = readdir(LIBDIR))) {
	    $foundlib++ if /^libssl/;
	    $foundlib++ if /^libcrypto/;
	    $foundlib++ if /^ssleay32/;
	    $foundlib++ if /^libeay32/;
	}
	closedir(LIBDIR);
    }
    warn "$dir/lib does not seem to contain the SSLeay library files\n"
	unless $foundlib;

    [$dir, $inc_dir, $version];
}


# end of more Crypt-SSLeay



MHTTP.pm
package HTTP::MHTTP;
use strict;
require 5.005;
use Carp;
require DynaLoader;
require Exporter;
use MIME::Base64 qw(encode_base64);
use vars qw(@ISA $VERSION @EXPORT_OK);
$VERSION = '0.15';
@ISA = qw(DynaLoader Exporter);


sub dl_load_flags { 0x01 }
HTTP::MHTTP->bootstrap($VERSION);


#  the supported request headers
  my $headers = {
                   'Accept-Encoding' => '0',
                   'Accept-Language' => '1',
                   'Connection'      => '2',
                   'Cookie'          => '3',
                   'Host'            => '4',
                   'User-Agent'      => '5',
                   'Authorization'   => '6',
                   'Accept'          => '7',
                   'SOAPAction'      => '8',
                   'Content-Type'    => '9',
                   'Cache-control'   => '10',
                   'Cache-Control'   => '10',
                   'Accept-Charset'  => '11',
                   'Pragma'          => '12',
                   'Referrer'        => '13',
                   'Referer'         => '13',
                   'Keep-Alive'      => '14',
                   'If-Modified-Since' => '15',
                   'Content-type'    => '16',
		 };



=head1 NAME

HTTP::MHTTP - low-level library for making HTTP requests

=head1 SYNOPSIS

 use HTTP::MHTTP;
  
 http_init();
  
 http_add_headers(
               'User-Agent' => 'DVSGHTTP1/1',
               'Accept-Language' => 'en-gb',
               'Connection' => 'Keep-Alive',
                   );
 if (http_call("GET", "http://localhost") > 0) {
   if (http_status() == 200 ) {
     print http_response();
   } else {
     print "MSG: ".http_reason();
   }
 } else {
   print "call failed \n";
 }

=head1 DESCRIPTION

This library provides reasonably low level access to the HTTP protocol, for perl.
This does not replace LWP (what possibly could :-) but is a cut for speed.
It also supports all of HTTP 1.0, so you have GET, POST, PUT, HEAD, and DELETE.
Some support of HTTP 1.1 is available - sepcifically Transfer-Encoding = chunked and the Keep-Alive extensions.

Additionally - rudimentary SSL support can be compiled in.
This effectively enables negotiation of TLS,
but does not validate the certificates.

=head2 http_init()

Initialise the mhttp library - must be called once to reset all internals,
use http_reset() if you don't need to reset your headers before the next call.


=head2 http_set_protocol()

  http_set_protocol(1);  # now operating in HTTP 1.1 mode

Set the protocol level to use - either HTTP 1.0 or 1.1 by passing 0 or 1 - 
the default is 0 (HTTP 1.0).


=head2 http_reset()

Reset the library internals for everything except the headers specified 
previously, and the debug switch.  Call http_init() if you need to reset
everything.


=head2 switch_debug()

  switch_debug(<0 || 1>)

Toggle the internal debugging on and off by passing either > 1 or 0.


=head2 http_add_headers()

  http_add_headers(
                 'User-Agent' => 'HTTP-MHTTP1/0',
                 'Host' => 'localhost',
                 'Accept-Language' => 'en-gb',
                );

pass in header/value pairs that will be set on the next http_call().


=head2 http_body()

  http_body("this is the body");

Set the body of the next request via http_call().


=head2 http_call()

  my $rc = http_call("GET", "http://localhost");

Do an http request.  Returns either < 0 or 1 depending on whether the call was 
successful - remember to still check the http_status() code though.

Value < 0 are:

    -1 : an invalid action (HTTP verb) was supplied
    -2 : must supply an action (HTTP verb)
    -3 : must supply a url
    -4 : url must start with http:// or https://
    -5 : write of headers to socket failed
    -6 : write of data to socket was short
    -7 : failed to write last line to socket
    -8 : something wrong with the Conent-Length header
   -11 : SSL_CTX_new failed - abort everything
   -12 : SSL_new failed - abort everything
   -13 : SSL_connect failed - abort everything
   -14 : SSL_get_peer_certificate failed - abort everything
   -15 : X509_get_subject_name failed - abort everything
   -16 : X509_get_issuer_name failed - abort everything
   -17 : cant find the next chunk for Transfer-encoding
   -18 : cant find end headers
   -19 : You must supply a Host header for HTTP/1.1


=head2 http_status()

Returns the last status code.


=head2 http_reason()

Returns the last reason code.


=head2 http_headers()

Returns the headers of the last call, as a single string.


=head2 http_split_headers()

Returns the split out array ref of array ref header value pairs of the last call. 
[ [ hdr, val], [hdr, val] ... ]


=head2 http_response_length()

Returns the length of the body of the last call.


=head2 http_response()

Returns the body of the last call.


=head2 basic_authorization()

  my $pass = basic_authorization($user, $password);

Construct the basic authorization value to be passed in an "Authorization"
header.


=head1 COPYRIGHT

Copyright (c) 2003, Piers Harding. All Rights Reserved.
This module is free software. It may be used, redistributed
and/or modified under the same terms as Perl itself.

=head1 AUTHOR

Piers Harding, piers@ompa.net.

=head1 SEE ALSO

perl(1)

=cut


# export the open command, and initialise http::mhttp
my @export_ok = ("http_reset", "http_init", "http_add_headers", "http_status", "http_reason", "http_call", "http_headers", "http_split_headers", "http_body", "http_response", "basic_authorization", "switch_debug", "http_response_length", "http_set_protocol" );
sub import {

  my ( $caller ) = caller;

  my ($me, $debug) = @_;

  no strict 'refs';
  foreach my $sub ( @export_ok ){
    *{"${caller}::${sub}"} = \&{$sub};
  }

}


sub http_add_headers {
  my $hdrs = { @_ };
  foreach my $header ( keys %$hdrs ){
    if ( exists $headers->{$header} ){
      add_header($header.": ".$hdrs->{$header});
    } else {
      warn "Invalid header specified: $header - $hdrs->{$header} \n";
    }
  }
}


sub http_split_headers {

  my $headers = [];
  foreach my $h (split(/\n/,http_headers())){
    next unless $h =~ /:/;
    my ($hdr,$val) = $h =~ /^(.*?):\s(.*?)$/;
    $val =~ s/[\n\r]//g;
    push (@$headers, [$hdr, $val]);
    #$headers->{$hdr} = $val;
  }
  return $headers;

}


sub basic_authorization{
  my ( $user, $passwd ) = @_;
  return "Basic ".encode_base64( $user.':'.$passwd, "" );
}

1;
